daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.speedy
+
+/**
+  Transformation to ANF based AST for the speedy interpreter.
+
+  "ANF" stands for A-normal form.
+  In essence it means that sub-expressions of most expression nodes are in atomic-form.
+  The one exception is the let-expression.
+
+  Atomic mean: a variable reference (ELoc), a (literal) value, or a builtin.
+  This is captured by any speedy-expression which `extends SExprAtomic`.
+
+  TODO: <EXAMPLE HERE>
+
+  The reason we convert to ANF is to improve the efficiency of speedy execution: the
+  execution engine can take advantage of the atomic assumption, and often removes
+  additional execution steps - in particular the pushing of continuations to allow
+  execution to continue after a compound expression is reduced to a value.
+
+  The speedy machine now expects that it will never have to execute a non-ANF expression,
+  crashing at runtime if one is encountered.  In particular we must ensure that the
+  expression forms: SEAppGeneral and SECase are removed, and replaced by the simpler
+  SEAppAtomic and SECaseAtomic (plus SELet as required).
+
+  */
+import com.daml.lf.speedy.SExpr._
+
+import scala.annotation.tailrec
+
+object Anf {
+
+  /*** Entry point for the ANF transformation phase */
+  def flattenToAnf(exp: SExpr): AExpr = {
+    val depth = DepthA(0)
+    val env = initEnv
+    flattenExp(depth, env, exp, flattenedExpression => Land(flattenedExpression)).bounce
+  }
+
+  /**
+    The transformation code is implemented using a continuation-passing style of
+    translation (which is quite common for translation to ANF). In general, naming nested
+    compound expressions requires turning an expression kind of inside out, lifting the
+    introduced let-expression up to the the nearest enclosing abstraction or case-branch.
+
+    For speedy, the ANF pass occurs after translation to De Brujin and closure conversions,
+    which adds the additional complication of re-indexing the variable indexes. This is
+    achieved by tracking the old and new depth & the mapping between them. See the types:
+    DepthE, DepthA and Env.
+
+    There is also the issue of avoiding stack-overflow during compilation, which is
+    managed by the using of a Trampoline[T] type.
+    */
+  case class CompilationError(error: String) extends RuntimeException(error)
+
+  /** `DepthE` tracks the stack-depth of the original expression being traversed */
+  case class DepthE(n: Int)
+
+  /** `DepthA` tracks the stack-depth of the ANF expression being constructed */
+  case class DepthA(n: Int)
```

Comment:
The code below might become less verbose if we use something like
```suggestion
  case class DepthA(n: Int) {
    def incr(m: Int) = DepthA(n+m)
  }
```
This might also help for `DepthE`.

Reply:
I made the change for `DepthA` and `DepthE`. It's nice. Thanks.